File: website/src/main/java/info/xiaomo/website/service/impl/ChangeLogServiceImpl.java
package info.xiaomo.website.service.impl;

import info.xiaomo.website.dao.ChangeLogDao;
import info.xiaomo.website.model.ChangeLogModel;
import info.xiaomo.website.service.ChangeLogService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;

import java.util.Date;
import java.util.List;
import java.util.Optional;

/**
 * 把今天最好的表现当作明天最新的起点．．～
 * いま 最高の表現 として 明日最新の始発．．～
 * Today the best performance  as tomorrow newest starter!

 *
 * @author : xiaomo
 * github: https://github.com/houko
 * email: xiaomo@xiaomo.info
 * <p>
 * Date: 2016/4/11 19:49
 * Copyright(©) 2015 by xiaomo.
 **/
@Service
public class ChangeLogServiceImpl implements ChangeLogService {

    private final ChangeLogDao dao;

    @Autowired
    public ChangeLogServiceImpl(ChangeLogDao dao) {
        this.dao = dao;
    }

    @Override
    public ChangeLogModel findById(Long id) {
        Optional<ChangeLogModel> changeLogModel = dao.findById(id);
        return changeLogModel.orElse(null);
    }

    @Override
    public ChangeLogModel findByName(String name) {
        return dao.findByName(name);
    }

    @Override
    public Page<ChangeLogModel> findAll(int start, int pageSize) {
        Sort sort = Sort.by(Sort.Direction.DESC, "createTime");
        return dao.findAll(PageRequest.of(start - 1, pageSize, sort));
    }

    @Override
    public List<ChangeLogModel> findAll() {
        return dao.findAll();
    }

    @Override
    public ChangeLogModel add(ChangeLogModel model) {
        model.setCreateTime(new Date());
        model.setUpdateTime(new Date());
        return dao.save(model);
    }

    @Override
    public ChangeLogModel update(ChangeLogModel model) {
        Optional<ChangeLogModel> optionalModel = dao.findById(model.getId());
        if (!optionalModel.isPresent()) {
            return null;
        }
        ChangeLogModel changeLogModel = optionalModel.get();
        if (model.getName() != null) {
            changeLogModel.setName(model.getName());
        }
        changeLogModel.setUpdateTime(new Date());
        return dao.save(changeLogModel);
    }

    @Override
    public ChangeLogModel delete(Long id) {
        Optional<ChangeLogModel> optionalLogModel = dao.findById(id);
        if (!optionalLogModel.isPresent()) {
            return null;
        }
        ChangeLogModel changeLogModel = optionalLogModel.get();
        dao.delete(changeLogModel);
        return changeLogModel;
    }
}

File: website/src/main/java/info/xiaomo/website/service/impl/ChangeLogServiceImpl.kt
package info.xiaomo.website.service.impl

import info.xiaomo.website.dao.ChangeLogDao
import info.xiaomo.website.model.ChangeLogModel
import info.xiaomo.website.service.ChangeLogService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import java.util.*

/**
 * 把今天最好的表现当作明天最新的起点．．～
 * いま 最高の表現 として 明日最新の始発．．～
 * Today the best performance  as tomorrow newest starter!
 * Created by IntelliJ IDEA.
 *
 * @author : xiaomo
 * github: https://github.com/xiaomoinfo
 * email: xiaomo@xiaomo.info
 *
 * Date: 2016/4/11 19:49
 * Copyright(©) 2015 by xiaomo.
 */
@Service
class ChangeLogServiceImpl @Autowired
constructor(private val dao: ChangeLogDao) : ChangeLogService {

    override fun findById(id: Long?): ChangeLogModel {
        return dao.findOne(id)
    }

    override fun findByName(name: String): ChangeLogModel {
        return dao.findByName(name)
    }

    override fun findAll(start: Int, pageSize: Int): Page<ChangeLogModel> {
        val sort = Sort(Sort.Direction.DESC, "createTime")
        return dao.findAll(PageRequest(start - 1, pageSize, sort))
    }

    override fun findAll(): List<ChangeLogModel> {
        return dao.findAll()
    }

    override fun add(model: ChangeLogModel): ChangeLogModel {
        model.createTime = Date()
        model.updateTime = Date()
        return dao.save(model)
    }

    override fun update(model: ChangeLogModel): ChangeLogModel {
        val updateModel = dao.findOne(model.id)
        if (model.name != null) {
            updateModel.name = model.name
        }
        updateModel.updateTime = Date()
        return dao.save(updateModel)
    }

    override fun delete(id: Long?): ChangeLogModel {
        val model = dao.findOne(id)
        if (model != null) {
            dao.delete(id)
        }
        return model
    }
}
